File: android-beans/src/main/java/rocks/frieler/android/beans/Beans.java
package rocks.frieler.android.beans;

import java.util.List;

import rocks.frieler.android.beans.scopes.ScopedFactoryBeanHandler;

/**
 * Static facade-class to access the beans.
 */
public final class Beans {
    private static BeansProvider beansProvider;

    static void setBeans(BeansProvider beansProvider) {
        Beans.beansProvider = beansProvider;
    }

    /**
     * Looks up the bean with the given name and type in the {@link BeansProvider} of this application.
     *
     * @param name the name of the desired bean
     * @param type the type of the desired bean
     * @param <T> the bean-type
     * @return the named bean or {@code null}
     *
     * @see BeansProvider#lookUpBean(String, Class)
     */
    public static <T> T lookUpBean(String name, Class<T> type) {
        return beansProvider.lookUpBean(name, type);
    }

    /**
     * Looks up a bean of the given type in the {@link BeansProvider} of this application.
     *
     * @param type the type of the desired bean
     * @param <T> the bean-type
     * @return a bean of the given type or {@code null}
     *
     * @see BeansProvider#lookUpBean(Class)
     */
    public static <T> T lookUpBean(Class<T> type) {
        return beansProvider.lookUpBean(type);
    }

    /**
     * Looks up all beans of the given type in the {@link BeansProvider} of this application.
     *
     * @param type the type of the desired beans
     * @param <T> the bean-type
     * @return the beans of the given type or an empty list
     *
     * @see BeansProvider#lookUpBeans(Class)
     */
    public static <T> List<T> lookUpBeans(Class<T> type) {
        return beansProvider.lookUpBeans(type);
    }

    private Beans() {
    }

    /**
     * Initializer for {@link Beans}.
     * <p>
     * Initializes {@link Beans} with a {@link BeanRegistry} and allows to configure it.
     */
    public static final class Initializer {
        private BeanRegistry beanRegistry = new BeanRegistry();

        /**
         * Adds a bean-scope expressed by the given {@link ScopedFactoryBeanHandler}.
         *
         * @param scopedFactoryBeanHandler the {@link ScopedFactoryBeanHandler} to add
         * @return the {@link Initializer} itself
         *
         * @see BeanRegistry#addBeanScope(ScopedFactoryBeanHandler)
         */
        public Initializer addScope(ScopedFactoryBeanHandler scopedFactoryBeanHandler) {
            beanRegistry.addBeanScope(scopedFactoryBeanHandler);
            return this;
        }

        /**
         * Collects beans from the given {@link BeanConfiguration}s.
         *
         * @param beanConfigurations the BeanConfigurations that define the beans
         * @return the {@link Initializer} itself
         *
         * @see BeanConfigurationsBeansCollector#collectBeans(List)
         */
        public Initializer collectBeans(List<? extends BeanConfiguration> beanConfigurations) {
            new BeanConfigurationsBeansCollector(beanRegistry).collectBeans(beanConfigurations);
            return this;
        }

        /**
         * Initializes {@link Beans} with the applied configuration.
         */
        public void initialize() {
            setBeans(beanRegistry);
        }
    }
}

File: android-beans/src/main/kotlin/rocks/frieler/android/beans/Beans.kt
package rocks.frieler.android.beans

import rocks.frieler.android.beans.scopes.ScopedFactoryBeanHandler
import kotlin.reflect.KClass

/**
 * Facade-object to access the beans in the global [BeansProvider].
 *
 * [Beans] needs to be initialized with the [Beans.Initializer] first.
 *
 * @author Christopher Frieler
 */
object Beans {
    private lateinit var beansProvider: BeansProvider

    internal fun setBeans(beansProvider: BeansProvider) {
        Beans.beansProvider = beansProvider
    }

    /**
     * Looks up the bean with the inferred type and given name (if not `null`) in the
     * [BeansProvider] of this application.
     *
     * @param name the name of the desired bean (optional)
     * @param <T> the desired bean-type
     * @return the bean
     *
     * @see lookUpBean
     */
    inline fun <reified T : Any> lookUpBean(name: String? = null) = lookUpBean(name, T::class)

    /**
     * Looks up the bean with the given name (if not `null`) and type in the [BeansProvider] of this
     * application.
     *
     * @param name the name of the desired bean (optional)
     * @param type the type of the desired bean
     * @param <T> the bean-type
     * @return the bean
     *
     * @see BeansProvider.lookUpBean
     */
    fun <T : Any> lookUpBean(name: String? = null, type: KClass<T>): T {
        return if (name == null)
            beansProvider.lookUpBean(type)
        else
            beansProvider.lookUpBean(name, type)
    }

    /**
     * Looks up the bean with the given name (if not `null`) and type in the [BeansProvider] of this
     * application.
     *
     * @param name the name of the desired bean (optional)
     * @param type the type of the desired bean
     * @param <T> the bean-type
     * @return the named bean
     *
     * @see lookUpBean
     */
    @JvmStatic
    @JvmOverloads
    fun <T :Any> lookUpBean(name: String? = null, type: Class<T>) = lookUpBean(name, type.kotlin)

    /**
     * Looks up the bean with the inferred type and given name (if not `null`) in the
     * [BeansProvider] of this application.
     *
     * @param name the name of the desired bean (optional)
     * @param <T> the desired bean-type
     * @return the bean or `null`
     *
     * @see lookUpOptionalBean
     */
    inline fun <reified T : Any> lookUpOptionalBean(name: String? = null) = lookUpOptionalBean(name, T::class)

    /**
     * Looks up the bean with the given name (if not `null`) and type in the [BeansProvider] of this
     * application.
     *
     * @param name the name of the desired bean (optional)
     * @param type the type of the desired bean
     * @param <T> the bean-type
     * @return the bean or `null`
     *
     * @see BeansProvider.lookUpOptionalBean
     */
    fun <T : Any> lookUpOptionalBean(name: String? = null, type: KClass<T>): T? {
        return if (name == null)
            beansProvider.lookUpOptionalBean(type)
        else
            beansProvider.lookUpOptionalBean(name, type)
    }

    /**
     * Looks up the bean with the given name (if not `null`) and type in the [BeansProvider] of this
     * application.
     *
     * @param name the name of the desired bean (optional)
     * @param type the type of the desired bean
     * @param <T> the bean-type
     * @return the named bean or `null`
     *
     * @see lookUpOptionalBean
     */
    @JvmStatic
    @JvmOverloads
    fun <T :Any> lookUpOptionalBean(name: String? = null, type: Class<T>) = lookUpOptionalBean(name, type.kotlin)

    /**
     * Looks up all beans of the inferred type in the [BeansProvider] of this application.
     *
     * @param <T> the desired bean-type
     * @return the beans of the given type or an empty list
     *
     * @see BeansProvider.lookUpBeans
     */
    inline fun <reified T : Any> lookUpBeans() = lookUpBeans(T::class)

    /**
     * Looks up all beans of the given type in the [BeansProvider] of this application.
     *
     * @param type the type of the desired beans
     * @param <T> the bean-type
     * @return the beans of the given type or an empty list
     *
     * @see BeansProvider.lookUpBeans
     */
    fun <T : Any> lookUpBeans(type: KClass<T>): List<T> {
        return beansProvider.lookUpBeans(type)
    }

    /**
     * Looks up all beans of the given type in the [BeansProvider] of this application.
     *
     * @param type the type of the desired beans
     * @param <T> the bean-type
     * @return the beans of the given type or an empty list
     *
     * @see lookUpBeans
     */
    @JvmStatic
    fun <T :Any> lookUpBeans(type: Class<T>) = lookUpBeans(type.kotlin)

    /**
     * Initializer for [Beans].
     *
     * Initializes [Beans] with a [BeanRegistry] and allows to configure it.
     */
    class Initializer {
        private val beanRegistry = BeanRegistry()

        /**
         * Adds a bean-scope expressed by the given [ScopedFactoryBeanHandler].
         *
         * @param scopedFactoryBeanHandler the [ScopedFactoryBeanHandler] to add
         * @return the [Initializer] itself
         *
         * @see BeanRegistry.addBeanScope
         */
        fun addScope(scopedFactoryBeanHandler: ScopedFactoryBeanHandler): Initializer {
            beanRegistry.addBeanScope(scopedFactoryBeanHandler)
            return this
        }

        /**
         * Collects beans from the given [BeanConfiguration]s.
         *
         * @param beanConfigurations the BeanConfigurations that define the beans
         * @return the [Initializer] itself
         *
         * @see BeanConfigurationsBeansCollector.collectBeans
         */
        fun collectBeans(beanConfigurations: List<BeanConfiguration>): Initializer {
            BeanConfigurationsBeansCollector(beanRegistry).collectBeans(beanConfigurations)
            return this
        }

        /**
         * Initializes [Beans] with the applied configuration.
         */
        fun initialize() {
            setBeans(beanRegistry)
        }
    }
}
